extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/BasicRoleBasedAuthorizer.java

```diff
@@ -0,0 +1,142 @@
+
+import com.fasterxml.jackson.annotation.JacksonInject;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import io.druid.java.util.common.IAE;
+import io.druid.security.basic.db.BasicAuthDBConfig;
+import io.druid.security.basic.db.BasicAuthorizerStorageConnector;
+import io.druid.server.security.Access;
+import io.druid.server.security.Action;
+import io.druid.server.security.AuthenticationResult;
+import io.druid.server.security.Authorizer;
+import io.druid.server.security.Resource;
+import io.druid.server.security.ResourceAction;
+
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@JsonTypeName("basic")
+public class BasicRoleBasedAuthorizer implements Authorizer
+{
+  private static final int DEFAULT_PERMISSION_CACHE_SIZE = 5000;
+  private final BasicAuthorizerStorageConnector dbConnector;
+  private final LoadingCache<String, Pattern> permissionPatternCache;
+  private final int permissionCacheSize;
+  private final BasicAuthDBConfig dbConfig;
+
+  @JsonCreator
+  public BasicRoleBasedAuthorizer(
+      @JacksonInject BasicAuthorizerStorageConnector dbConnector,
+      @JsonProperty("dbPrefix") String dbPrefix,
+      @JsonProperty("permissionCacheSize") Integer permissionCacheSize
+  )
+  {
+    this.dbConfig = new BasicAuthDBConfig(dbPrefix, null, null);
+    this.dbConnector = dbConnector;
+    this.permissionCacheSize = permissionCacheSize == null ? DEFAULT_PERMISSION_CACHE_SIZE : permissionCacheSize;
+    this.permissionPatternCache = Caffeine.newBuilder()
+                                          .maximumSize(this.permissionCacheSize)
+                                          .build(regexStr -> Pattern.compile(regexStr));
+  }
+
+  /**
+   * constructor for unit tests
+   */
+  public BasicRoleBasedAuthorizer(
+      BasicAuthorizerStorageConnector dbConnector,
+      String dbPrefix,
+      int permissionCacheSize
+  )
+  {
+    this.dbConnector = dbConnector;
+    this.permissionCacheSize = permissionCacheSize;
+    this.permissionPatternCache = Caffeine.newBuilder()
+                                          .maximumSize(this.permissionCacheSize)
+                                          .build(regexStr -> Pattern.compile(regexStr));
+    this.dbConfig = new BasicAuthDBConfig(dbPrefix, null, null);
+  }
+
+  @Override
+  public Access authorize(
+      AuthenticationResult authenticationResult, Resource resource, Action action
+  )
+  {
+    if (authenticationResult == null) {
+      throw new IAE("WTF? authenticationResult should never be null.");
+    }
+
+    List<Map<String, Object>> permissions = dbConnector.getPermissionsForUser(dbConfig.getDbPrefix(), authenticationResult.getIdentity());
```

Comment:
We spoke about this offline, but, it's important that we don't make a DB call in the hot path of Druid queries, for a couple reasons:

- DB calls take time and we want Druid queries to be as fast as possible
- The DB can be down and we want Druid queries to still function during DB outages

So for this reason, all necessary information to make a permissions decision should be cached, and updated in the background, never synchronously during a call to `authorize`. While doing this we should also make sure to avoid the issues that have come up with similar functionality in QTLs:

- we should sync permissions details on startup before serving queries, to avoid spurious rejections (similar to QTL #4758)
- we should make sure that permissions syncing has retries, to handle situations where permission storage is briefly down
- the retries should have backoff, to avoid DDoSing wherever the permissions are stored
- background polling should have some fuzz to avoid needless thundering (I'm not sure if QTLs do this one, but if they don't then they should!)

In addition, IMO it's best for the coordinator to be the only thing that actually retrieves permissions from the metadata store -- and for the rest of the nodes to hit the coordinator to get them from the coordinator's cache. Imagine a 500 node Druid cluster. The coordinator is better able to handle that load than the metadata store (which is often relatively small compared to the coordinator).